tools/clang/unittests/HLSLExec/LongVectors.cpp

```diff
@@ -1013,6 +1014,49 @@ BINARY_COMPARISON_OP(OpType::NotEqual, (A != B));
 DEFAULT_OP_2(OpType::Logical_And, (A && B));
 DEFAULT_OP_2(OpType::Logical_Or, (A || B));
 
+//
+// Reduction
+//
+
+#define REDUCTION_OP(OP, STDFUNC)                                              \
+  template <typename T> struct Op<OP, T> : StrictValidation {};                \
+  template <typename T> struct ExpectedBuilder<OP, T> {                        \
+    static std::vector<HLSLBool_t>                                             \
+    buildExpected(Op<OP, T>, const InputSets<T, 1> &Inputs, uint16_t) {        \
+      const bool Res = STDFUNC(Inputs[0].begin(), Inputs[0].end(),             \
+                               [](T A) { return A != static_cast<T>(0); });    \
+      return std::vector<HLSLBool_t>{Res};                                     \
+    }                                                                          \
+  };
+
+REDUCTION_OP(OpType::Any_Mixed, (std::any_of));
+REDUCTION_OP(OpType::Any_NoZero, (std::any_of));
+REDUCTION_OP(OpType::Any_Zero, (std::any_of));
+
+REDUCTION_OP(OpType::All_Mixed, (std::all_of));
+REDUCTION_OP(OpType::All_NoZero, (std::all_of));
+REDUCTION_OP(OpType::All_Zero, (std::all_of));
+
+#undef REDUCTION_OP
+
+template <typename T> struct Op<OpType::Dot, T> : DefaultValidation<T> {};
+template <typename T> struct ExpectedBuilder<OpType::Dot, T> {
+  static std::vector<T> buildExpected(Op<OpType::Dot, T>,
+                                      const InputSets<T, 2> &Inputs,
+                                      uint16_t ScalarInputFlags) {
+    T DotProduct = T();
+
+    for (size_t I = 0; I < Inputs[0].size(); ++I) {
```

Comment:
I am having a very hard time parsing what this loop is actually doing. From context its calculating the dot product but just by reading the code I can't seem to tease that out.

I think part of that is trying to account for scalars? But honestly it feels a bit too clever. Can we just pull the first part out of the loop?

Reply:
I guess it doesn't hurt to remove that logic for this PR though as I've already excluded those test cases for Dot.

Reply:
I'll leave it up to you! Looks like there is already a merge conflict so you won't save on CI time by delaying but regardless, I'm happy either way